### PR TITLE
Added P42 ver2 with Feather support. Updated keymaps.

### PR DESCRIPTION
**Pull request by pabile**
Separated promicro from feather for P42 keyboard. Added keymaps for P20 and P4w keyboards

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [ ] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [ ] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/pabile/p42/feather/keymaps/default/keymap.c [hunk 1]
```diff
@@ -0,0 +1,30 @@
+#include QMK_KEYBOARD_H
```
Comment (drashna) [inline keyboards/pabile/p42/feather/keymaps/default/keymap.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/pabile/p42/promicro/ver2/keymaps/default/keymap.c [hunk 2]
```diff
@@ -0,0 +1,29 @@
+#include QMK_KEYBOARD_H
```
Comment (drashna) [inline keyboards/pabile/p42/promicro/ver2/keymaps/default/keymap.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/pabile/p42/feather/keymaps/default/keymap.c (hunk 1 above)
Comment (pabile) [reply keyboards/pabile/p42/feather/keymaps/default/keymap.c:19]: thanks for checking, drashna. changes made.

File: keyboards/pabile/p42/promicro/ver2/keymaps/default/keymap.c (hunk 2 above)
Comment (pabile) [reply keyboards/pabile/p42/promicro/ver2/keymaps/default/keymap.c:19]: done, drashana. thanks

File: keyboards/pabile/p20/ver2/keymaps/backlight/keymap.c [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+void encoder_update_user(uint8_t index, bool clockwise) {
+  if (index == 0) { /* First encoder */
+    if (clockwise) {
+      backlight_decrease();
+    } else {
+      backlight_increase();
+    }
+  } else if (index == 1) { /* Second encoder */  
+    if (clockwise) {
+     SEND_STRING(SS_LCTL("-")); /* zoom in */
```
Comment (fauxpark) [inline keyboards/pabile/p20/ver2/keymaps/backlight/keymap.c:54]:
```suggestion
     tap_code16(C(KC_MINS)); /* zoom in */
```

File: keyboards/pabile/p20/ver2/keymaps/backlight/keymap.c [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    } else { 
+     SEND_STRING(SS_LCTL("=")); /* zoom out */
```
Comment (fauxpark) [inline keyboards/pabile/p20/ver2/keymaps/backlight/keymap.c:56]:
```suggestion
     tap_code16(C(KC_EQL)); /* zoom out */
```

File: keyboards/pabile/p20/ver2/keymaps/4encoders/keymap.c [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+void encoder_update_user(uint8_t index, bool clockwise) {
+  if (index == 0) { /* First encoder */
+    if (clockwise) {
+      tap_code(KC_LBRC); /* brush size small */
+    } else {
+      tap_code(KC_RBRC); /* brush size big */
+    }
+  } else if (index == 1) { /* Second encoder */  
+    if (clockwise) {
+     SEND_STRING(SS_LCTL("-")); /* zoom in */
```
Comment (fauxpark) [inline keyboards/pabile/p20/ver2/keymaps/4encoders/keymap.c:54]:
```suggestion
     tap_code16(C(KC_MINS)); /* zoom in */
```

File: keyboards/pabile/p20/ver2/keymaps/4encoders/keymap.c [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    } else { 
+     SEND_STRING(SS_LCTL("=")); /* zoom out */
```
Comment (fauxpark) [inline keyboards/pabile/p20/ver2/keymaps/4encoders/keymap.c:56]:
```suggestion
     tap_code16(C(KC_EQL)); /* zoom out */
```

File: keyboards/pabile/p20/ver2/keymaps/via4encoders/keymap.c [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+void encoder_update_user(uint8_t index, bool clockwise) {
+  if (index == 0) { /* First encoder */
+    if (clockwise) {
+      tap_code(KC_LBRC); /* brush size small */
+    } else {
+      tap_code(KC_RBRC); /* brush size big */
+    }
+  } else if (index == 1) { /* Second encoder */  
+    if (clockwise) {
+     SEND_STRING(SS_LCTL("-")); /* zoom in */
```
Comment (fauxpark) [inline keyboards/pabile/p20/ver2/keymaps/via4encoders/keymap.c:54]:
```suggestion
     tap_code16(C(KC_MINS)); /* zoom in */
```

File: keyboards/pabile/p20/ver2/keymaps/via4encoders/keymap.c [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    } else { 
+     SEND_STRING(SS_LCTL("=")); /* zoom out */
```
Comment (fauxpark) [inline keyboards/pabile/p20/ver2/keymaps/via4encoders/keymap.c:56]:
```suggestion
     tap_code16(C(KC_EQL)); /* zoom out */
```

File: keyboards/pabile/p20/ver2/rules.mk [hunk 9]
```diff
@@ -7,7 +7,8 @@ BOOTLOADER = caterina
 # Build Options
 #   change yes to no to disable
 #
-BOOTMAGIC_ENABLE = no       # Virtual DIP switch configuration
+
+BOOTMAGIC_ENABLE = lite       # Virtual DIP switch configuration
```
Comment (fauxpark) [inline keyboards/pabile/p20/ver2/rules.mk:11]:
```suggestion
BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
```

File: keyboards/pabile/p20/ver2/readme.md [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+![Linear grid](https://github.com/pabile/Pabile20/blob/master/_bak/layout-grid.jpg)
+
+A 5x4 grid macro numeric pad with up to 3 optional rotary encoders. 
+
+* Keyboard Maintainer: [pabile](https://github.com/pabile)
+* Hardware Supported: Pabile P20 PCB, Pro Micro, Rotary Encoders, Alps/MX Switches
+* Hardware Availability: [pabileonline.blogspot.com](https://pabileonline.blogspot.com/search/label/p20)
+
+Make example for this keyboard (after setting up your build environment):
+    make pabile/p20/ver2:default
```
Comment (fauxpark) [inline keyboards/pabile/p20/ver2/readme.md:12]:
```suggestion

    make pabile/p20/ver2:default
```

File: keyboards/pabile/p40_ortho/rules.mk [hunk 11]
```diff
@@ -7,7 +7,8 @@ BOOTLOADER = caterina
 # Build Options
 #   change yes to no to disable
 #
-BOOTMAGIC_ENABLE = no       # Virtual DIP switch configuration
+
+BOOTMAGIC_ENABLE = lite       # Virtual DIP switch configuration
```
Comment (fauxpark) [inline keyboards/pabile/p40_ortho/rules.mk:11]:
```suggestion
BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
```

File: keyboards/pabile/p42/feather/rules.mk [hunk 12]
```diff
@@ -0,0 +1,28 @@
+# MCU name
+MCU = atmega32u4
+
+F_CPU = 8000000
```
Comment (fauxpark) [inline keyboards/pabile/p42/feather/rules.mk:4]:
```suggestion
# Processor frequency
F_CPU = 8000000
```

File: keyboards/pabile/p20/ver2/keymaps/backlight/config.h [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#define ENCODERS_PAD_B { D2, F4, F5, B3 }
+#define ENCODERS_PAD_A { D3, F7, B1, F6 }
+
+#define BACKLIGHT_PIN B6
+#define BACKLIGHT_BREATHING 1
```
Comment (fauxpark) [inline keyboards/pabile/p20/ver2/keymaps/backlight/config.h:24]:
```suggestion
#define BACKLIGHT_BREATHING
```

File: keyboards/pabile/p42/promicro/readme.md [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+* Hardware Availability: [pabileonline.blogspot.com](https://pabileonline.blogspot.com/)
+
+Make example for this keyboard (after setting up your build environment):
+For PCB ver1
+
+    make pabile/p42/promicro/ver1:default
+
+For PCB ver2 and above 
+
+    make pabile/p42/promicro/ver2:default
```
Comment (fauxpark) [inline keyboards/pabile/p42/promicro/readme.md:18]:
```suggestion

    make pabile/p42/promicro/ver1:default # For PCB ver1
    make pabile/p42/promicro/ver2:default # For PCB ver2 and above
```

File: keyboards/pabile/p42/promicro/ver1/rules.mk [hunk 15]
```diff
@@ -7,7 +7,7 @@ BOOTLOADER = caterina
 # Build Options
 #   change yes to no to disable
 #
-BOOTMAGIC_ENABLE = no       # Virtual DIP switch configuration
+BOOTMAGIC_ENABLE = lite       # Virtual DIP switch configuration
```
Comment (fauxpark) [inline keyboards/pabile/p42/promicro/ver1/rules.mk:10]:
```suggestion
BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
```

File: keyboards/pabile/p42/promicro/ver2/ver2.h.save [hunk 16]
```diff
@@ -0,0 +1,34 @@
+/*
```
Comment (fauxpark) [inline keyboards/pabile/p42/promicro/ver2/ver2.h.save:1]: Please delete this file.

File: keyboards/pabile/p42/readme.md [hunk 17]
```diff
@@ -5,11 +5,16 @@
 42-key Ortho-linear keyboard
 
 * Keyboard Maintainer: [pabile](https://github.com/pabile)
-* Hardware Supported: Pabile P42 PCB, Pro Micro, Bluefruit LE, Hotswap Sockets
+* Hardware Supported: Pabile P42 PCB, Pro Micro and Kailh Hotswap Sockets
 * Hardware Availability: [pabileonline.blogspot.com](https://pabileonline.blogspot.com/)
 
 Make example for this keyboard (after setting up your build environment):
+Pro Micro
```
Comment (fauxpark) [inline keyboards/pabile/p42/readme.md:12]:
```suggestion
```

File: keyboards/pabile/p42/readme.md [hunk 18]
```diff
@@ -5,11 +5,16 @@
...
 
-    make pabile/p42:default
+    make pabile/p42/promicro:default
+
+Feather 32U4 BLE
+  
+    make pabile/p42/feather:default
```
Comment (fauxpark) [inline keyboards/pabile/p42/readme.md:18]:
```suggestion
    make pabile/p42/promicro:default
    make pabile/p42/feather:default
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/pabile/p42/promicro/ver2/keymaps/default/keymap.c [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "ver2.h"
```
Comment (drashna) [inline keyboards/pabile/p42/promicro/ver2/keymaps/default/keymap.c:17]: This shouldn't be needed.  The qmk_keyboard_h should get this. 
```suggestion
```

File: keyboards/pabile/p42/feather/keymaps/via/keymap.c [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "feather.h"
```
Comment (drashna) [inline keyboards/pabile/p42/feather/keymaps/via/keymap.c:17]:
```suggestion
```

File: keyboards/pabile/p42/promicro/ver1/keymaps/1encoder/keymap.c [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "ver1.h"
```
Comment (drashna) [inline keyboards/pabile/p42/promicro/ver1/keymaps/1encoder/keymap.c:17]:
```suggestion
```

File: keyboards/pabile/p42/promicro/ver1/keymaps/default/keymap.c [hunk 22]
```diff
@@ -1,3 +1,21 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "ver1.h"
```
Comment (drashna) [inline keyboards/pabile/p42/promicro/ver1/keymaps/default/keymap.c:17]:
```suggestion
```

File: keyboards/pabile/p42/promicro/ver1/keymaps/overkill/keymap.c [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "ver1.h"
```
Comment (drashna) [inline keyboards/pabile/p42/promicro/ver1/keymaps/overkill/keymap.c:17]:
```suggestion
```

File: keyboards/pabile/p42/promicro/ver1/keymaps/via/keymap.c [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "ver1.h"
```
Comment (drashna) [inline keyboards/pabile/p42/promicro/ver1/keymaps/via/keymap.c:17]:
```suggestion
```

File: keyboards/pabile/p42/promicro/ver2/keymaps/via/keymap.c [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "ver2.h"
```
Comment (drashna) [inline keyboards/pabile/p42/promicro/ver2/keymaps/via/keymap.c:17]:
```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by pabile** [comment]
Hi @drahsna & @fauxpark 

I think this request got stuck and wanted to ask if there is anything that I should do my end? 

Thanks

**Comment by drashna** [comment]
There are still a few things that need addressing. However, they may be hidden by default. (thanks github).

**Comment by pabile** [comment]
> There are still a few things that need addressing. However, they may be hidden by default. (thanks github).

Thanks for the reply, @drashna I'll just wait for a notification if there is anything else needed on my side. Stay safe!

**Pull request closed without merging**